### ISSUE TITLE
uniteigenvectors fails if uv[1] is used in the user main program.

### ISSUE THREAD

**Issue by rtoy** [rtoy/maxima#3485]
Imported from SourceForge on 2024-07-08 01:55:20
Created by **[richardgobeli](https://sourceforge.net/u/richardgobeli/)** on 2020-09-15 01:39:41
Original: https://sourceforge.net/p/maxima/bugs/3654

---

load the program below and see that the uv array causes the uniteigenvectors function to fail even though the uv used by uniteigenvectors is define inside a block as a local variable.
```{maxima}
load(eigen);

m:matrix(
		[-(r-2·m)/r,	sqrt(2·m/r),	0,	0],
		[sqrt(2·m/r),	1,	0,	0],
		[0,	0,	r^2,	0],
		[0,	0,	0,	r^2·sin(θ)^2]
	);

uniteigenvectors(m);  /* function works here */

uv[1]:0;

uniteigenvectors(m);  /* fails here because the uv is not hidden in the function uniteigenvectors */
```

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-08 01:55:23
Created by **[richardgobeli](https://sourceforge.net/u/richardgobeli/)** on 2020-09-15 02:43:47
Original: https://sourceforge.net/p/maxima/bugs/3654/#c60d

---

The matrix should be named mm because there is a m in it, but that does not cause this fail.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-08 01:55:30
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2020-09-20 00:21:13
Original: https://sourceforge.net/p/maxima/bugs/3654/#bc35

---

Well, I don't know why the formatting didn't work. Anyway it looks OK now.

About this problem, it is combination of two bits of Maxima strangeness. One is the lack of lexical scope, so the symbol `uv` in the function is the same symbol as `uv` in the global level. The other is that if there is both a value and an array associated with a symbol (this is possible because Maxima is a so-called "Lisp-2"), then the array is referenced, i.e., `uv[2]` refers to the global `uv` array instead of the local value of `uv`.

The right way to fix this is to implement lexical scope, which I'm working on from time to time. There are a couple of work arounds. One is to rename local variables to something unusual. The other is to write `(+uv)[2]` instead of `uv[2]` -- this defeats the detection of the global array. I will give the second workaround a try, stay tuned.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-08 01:55:33
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2021-09-18 20:50:43
Original: https://sourceforge.net/p/maxima/bugs/3654/#5325

---

- **status**: open --> closed

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-08 01:55:37
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2021-10-07 15:37:22
Original: https://sourceforge.net/p/maxima/bugs/3654/#ead8

---

Fixed a different way by commit ee8cb0a. From the commit message:

> Fix SF bug #3654 in eigen.mac a different way (via local instead of (+foo)[i] workaround), essentially reverting commit a58e06 to eigen.mac and fixing it a different way.
> 
> This corrects a problem reported to the mailing list circa 2021-09-28, that running run_testsuite(share_tests = true) reports that rtest_eigen runs into a Maxima error ("cannot assign to" subscripted expression).
> 
> So the immediate problem is fixed, but the underlying problem is that rules applied to MQAPPLY (as in rtest6 circa line 82) change the way MQAPPLY expressions are simplified. I have opened a bug report, SF #3870: "rules applied to MQAPPLY change default simplification"
